models/build.php

```diff
@@ -387,6 +393,117 @@ private function GetRelatedBuildId($which_build_criteria)
         return 0;
     }
 
+    /**
+     * Return the errors that have been resolved from this build.
+     * @todo This doesn't support getting resolved build errors across parent builds.
+     **/
+    public function GetResolvedBuildErrors($type)
+    {
+        // This returns an empty result if there was no previous build
+        return pdo_query(
+            'SELECT * FROM
+             (SELECT * FROM builderror
+              WHERE buildid=' . $this->GetPreviousBuildId() . ' AND type=' . $type . ') AS builderrora
+             LEFT JOIN
+             (SELECT crc32 AS crc32b FROM builderror
+              WHERE buildid=' . $this->Id . ' AND type=' . $type . ') AS builderrorb
+              ON builderrora.crc32=builderrorb.crc32b
+             WHERE builderrorb.crc32b IS NULL');
+    }
+
+    // Is this redundant naming? Should it just be GetErrors?
+    /**
+     * Get all the errors in a build, of a given type.
+     *
+     * If the build is a parent, it returns all errors of child builds
+     * as well as the subprojectid.
+     **/
+    public function GetBuildErrors($type, $extrasql)
+    {
+        // @todo Needs to be profiled
+        if ($this->IsParentBuild()) {
+            return pdo_query("SELECT sp2b.subprojectid, sp.name subprojectname, be.* FROM builderror be
+                              JOIN build AS b ON b.id = be.buildid
+                              JOIN subproject2build AS sp2b ON sp2b.buildid = be.buildid
+                              JOIN subproject AS sp ON sp.id = sp2b.subprojectid
```

Comment:
I'd be happy to help you profile this stuff.  As an easy first pass, could we use INNER JOIN or LEFT JOIN here?